File: src/data/ContentDefinitions.js
import { normalizeZoneId } from '../world/ConstructionVocabulary.js';

/**
 * Renderer-independent authored definitions used by validation and domain
 * consumers. Stable IDs are persistence contracts; labels and presentation
 * metadata may change without migrating saves.
 */

export const WORLD_BOUNDS = Object.freeze({
  minX: -190,
  maxX: 810,
  minY: -100,
  maxY: 2_000,
  minZ: -390,
  maxZ: 390
});

export const DISTRICT_DEFINITIONS = Object.freeze([
  Object.freeze({
    id: 'WEST_CORE',
    label: 'West Core',
    bounds: Object.freeze({ minX: -190, maxX: 105, minZ: -150, maxZ: 150 }),
    releaseScope: 'MVP'
  }),
  Object.freeze({
    id: 'CENTRAL_PARK',
    label: 'Central Park',
    bounds: Object.freeze({ minX: -105, maxX: -45, minZ: -105, maxZ: -45 }),
    releaseScope: 'MVP'
  }),
  Object.freeze({
    id: 'PRIMARY_BRIDGE_CORRIDOR',
    label: 'Primary Bridge Corridor',
    bounds: Object.freeze({ minX: 105, maxX: 330, minZ: -150, maxZ: 150 }),
    releaseScope: 'MVP'
  }),
  Object.freeze({
    id: 'EAST_CYBER_METROPOLIS',
    label: 'East Cyber Metropolis',
    bounds: Object.freeze({ minX: 330, maxX: 810, minZ: -390, maxZ: 390 }),
    releaseScope: 'POST_MVP'
  })
]);

const zone = (id, label, {
  aliases = [],
  color,
  happiness = 0,
  landValue = 0,
  kind = 'DEVELOPMENT',
  releaseScope = 'MVP'
} = {}) => Object.freeze({
  id,
  label,
  aliases: Object.freeze([...aliases]),
  color,
  happiness,
  landValue,
  kind,
  releaseScope
});

export const ZONE_DEFINITIONS = Object.freeze([
  zone('RESIDENTIAL', 'Residential', {
    aliases: ['RES'], color: 0x22c55e, happiness: 1.2, landValue: 1.5
  }),
  zone('COMMERCIAL', 'Commercial', {
    aliases: ['COM', 'OFFICE'], color: 0xd946ef, happiness: 0.3, landValue: 2.2
  }),
  zone('OPERATIONS', 'Operations', {
    aliases: ['OPS', 'IND', 'INDUSTRIAL'], color: 0xf97316, happiness: -1.5, landValue: -1
  }),
  zone('POWER_SERVICE', 'Legacy Power Parcel', {
    aliases: ['POWER'], color: 0xfacc15, happiness: 0.2, landValue: 0.4,
    kind: 'SERVICE', releaseScope: 'COMPATIBILITY'
  }),
  zone('WATER_SERVICE', 'Legacy Water Parcel', {
    aliases: ['WATER'], color: 0x06b6d4, happiness: 0.8, landValue: 0.7,
    kind: 'SERVICE', releaseScope: 'COMPATIBILITY'
  }),
  zone('FIRE_SERVICE', 'Legacy Fire Parcel', {
    aliases: ['FIRE'], color: 0xef4444, happiness: 1.5, landValue: 1.2,
    kind: 'SERVICE', releaseScope: 'COMPATIBILITY'
  }),
  zone('SUBURBAN_RESIDENTIAL', 'Suburban Residential', {
    color: 0x65a30d, happiness: 1.4, landValue: 1.1,
    kind: 'AUTHORED_WORLD', releaseScope: 'POST_MVP'
  })
]);

const ZONES_BY_INPUT = new Map();
for (const definition of ZONE_DEFINITIONS) {
  ZONES_BY_INPUT.set(definition.id, definition);
  for (const alias of definition.aliases) ZONES_BY_INPUT.set(alias, definition);
}

export function getZoneDefinition(idOrAlias) {
  if (typeof idOrAlias !== 'string') return null;
  const normalized = normalizeZoneId(idOrAlias);
  return ZONES_BY_INPUT.get(normalized || idOrAlias.trim().toUpperCase()) || null;
}

export const FACTION_DEFINITIONS = Object.freeze([
  Object.freeze({ id: 'QUANTUM_DYNAMICS', label: 'Quantum Dynamics', minReputation: -100, maxReputation: 100 }),
  Object.freeze({ id: 'AETHER_SKYSPIRE', label: 'Aether Skyspire', minReputation: -100, maxReputation: 100 }),
  Object.freeze({ id: 'RESIDENTS', label: 'Residents', minReputation: -100, maxReputation: 100 }),
  Object.freeze({ id: 'OPERATIONS', label: 'Operations', minReputation: -100, maxReputation: 100 })
]);

export const PROGRESSION_DEFINITIONS = Object.freeze([
  Object.freeze({
    id: 'OPERATOR',
    label: 'Operator',
    rank: 1,
    prerequisiteIds: Object.freeze([])
  }),
  Object.freeze({
    id: 'BROKER',
    label: 'Broker',
    rank: 2,
    prerequisiteIds: Object.freeze(['OPERATOR'])
  }),
  Object.freeze({
    id: 'MAGNATE',
    label: 'Magnate',
    rank: 3,
    prerequisiteIds: Object.freeze(['BROKER'])
  })
]);

export const VEHICLE_CONTENT_IDS = Object.freeze([
  'SEDAN', 'SPORTS', 'SPORTS_CAR', 'BUS', 'TRUCK', 'TAXI', 'POLICE',
  'AMBULANCE', 'ICECREAM', 'DUMP_TRUCK', 'MOTORBIKE'
]);

File: godot/MetroPulse.Domain/Content/ContentDefinitions.cs
using System.Collections.Frozen;

namespace MetroPulse.Domain.Content;

public sealed record WorldBounds(double MinX, double MaxX, double MinY, double MaxY, double MinZ, double MaxZ)
{
    public bool Contains(double x, double z) => x >= MinX && x <= MaxX && z >= MinZ && z <= MaxZ;
}

public sealed record PlanarBounds(double MinX, double MaxX, double MinZ, double MaxZ);

public sealed record DistrictDefinition(string Id, string Label, PlanarBounds Bounds, string ReleaseScope);

public sealed record ZoneDefinition(
    string Id,
    string Label,
    IReadOnlyList<string> Aliases,
    int Color,
    double Happiness,
    double LandValue,
    string Kind,
    string ReleaseScope);

public sealed record FactionDefinition(string Id, string Label, double MinReputation, double MaxReputation);

public sealed record ProgressionDefinition(string Id, string Label, int Rank, IReadOnlyList<string> PrerequisiteIds);

public static class FeatureIds
{
    public const string Aircraft = "aircraft";
    public const string RocketLaunch = "rocketLaunch";
    public const string EastSideDevelopment = "eastSideDevelopment";
    public const string TemporaryMayhem = "temporaryMayhem";
    public const string MayhemVariants = "mayhemVariants";
    public const string PersistentMayhem = "persistentMayhem";
    public const string CountrysideExpansion = "countrysideExpansion";
}

public static class ContentDefinitions
{
    public static readonly WorldBounds WorldBounds = new(-190, 810, -100, 2_000, -390, 390);

    public static readonly IReadOnlyList<DistrictDefinition> Districts = Array.AsReadOnly<DistrictDefinition>([
        new("WEST_CORE", "West Core", new(-190, 105, -150, 150), "MVP"),
        new("CENTRAL_PARK", "Central Park", new(-105, -45, -105, -45), "MVP"),
        new("PRIMARY_BRIDGE_CORRIDOR", "Primary Bridge Corridor", new(105, 330, -150, 150), "MVP"),
        new("EAST_CYBER_METROPOLIS", "East Cyber Metropolis", new(330, 810, -390, 390), "POST_MVP"),
    ]);

    public static readonly IReadOnlyList<ZoneDefinition> Zones = Array.AsReadOnly([
        Zone("RESIDENTIAL", "Residential", ["RES"], 0x22c55e, 1.2, 1.5),
        Zone("COMMERCIAL", "Commercial", ["COM", "OFFICE"], 0xd946ef, 0.3, 2.2),
        Zone("OPERATIONS", "Operations", ["OPS", "IND", "INDUSTRIAL"], 0xf97316, -1.5, -1),
        Zone("POWER_SERVICE", "Legacy Power Parcel", ["POWER"], 0xfacc15, 0.2, 0.4, "SERVICE", "COMPATIBILITY"),
        Zone("WATER_SERVICE", "Legacy Water Parcel", ["WATER"], 0x06b6d4, 0.8, 0.7, "SERVICE", "COMPATIBILITY"),
        Zone("FIRE_SERVICE", "Legacy Fire Parcel", ["FIRE"], 0xef4444, 1.5, 1.2, "SERVICE", "COMPATIBILITY"),
        Zone("SUBURBAN_RESIDENTIAL", "Suburban Residential", [], 0x65a30d, 1.4, 1.1, "AUTHORED_WORLD", "POST_MVP"),
    ]);

    public static readonly IReadOnlyList<FactionDefinition> Factions = Array.AsReadOnly<FactionDefinition>([
        new("QUANTUM_DYNAMICS", "Quantum Dynamics", -100, 100),
        new("AETHER_SKYSPIRE", "Aether Skyspire", -100, 100),
        new("RESIDENTS", "Residents", -100, 100),
        new("OPERATIONS", "Operations", -100, 100),
    ]);

    public static readonly IReadOnlyList<ProgressionDefinition> Progression = Array.AsReadOnly<ProgressionDefinition>([
        new("OPERATOR", "Operator", 1, Array.AsReadOnly(Array.Empty<string>())),
        new("BROKER", "Broker", 2, Array.AsReadOnly(["OPERATOR"])),
        new("MAGNATE", "Magnate", 3, Array.AsReadOnly(["BROKER"])),
    ]);

    public static readonly IReadOnlyList<string> VehicleIds = Array.AsReadOnly([
        "SEDAN",
        "SPORTS",
        "SPORTS_CAR",
        "BUS",
        "TRUCK",
        "TAXI",
        "POLICE",
        "AMBULANCE",
        "ICECREAM",
        "DUMP_TRUCK",
        "MOTORBIKE",
    ]);

    public static readonly IReadOnlyList<string> MvpWorldFootprint = Array.AsReadOnly([
        "WEST_CORE",
        "CENTRAL_PARK",
        "PRIMARY_BRIDGE_CORRIDOR",
    ]);

    public static readonly IReadOnlyList<string> MvpActivityTemplates = Array.AsReadOnly([
        "TAXI",
        "COURIER",
        "DELIVERY",
        "RACE",
        "SABOTAGE",
        "SURVIVAL",
    ]);

    public static readonly IReadOnlyList<string> MvpMissionIds = Array.AsReadOnly([
        "mission_executive",
        "mission_scientist",
        "mission_police_robbery",
        "mission_police_park",
        "mission_sports_trial",
        "mission_sports_smuggle",
        "mission_bus_loop",
        "mission_truck_delivery",
        "mission_sedan_grocery",
        "mission_mayhem_escape",
    ]);

    public static readonly IReadOnlyDictionary<string, string> MvpZoneLabels =
        new Dictionary<string, string>
        {
            ["RESIDENTIAL"] = "Residential",
            ["COMMERCIAL"] = "Commercial",
            ["OPERATIONS"] = "Operations",
        }.ToFrozenDictionary(StringComparer.Ordinal);

    public static readonly IReadOnlyDictionary<string, bool> MvpFeatureFlags =
        new Dictionary<string, bool>
        {
            [FeatureIds.Aircraft] = false,
            [FeatureIds.RocketLaunch] = false,
            [FeatureIds.EastSideDevelopment] = false,
            [FeatureIds.TemporaryMayhem] = false,
            [FeatureIds.MayhemVariants] = false,
            [FeatureIds.PersistentMayhem] = false,
            [FeatureIds.CountrysideExpansion] = false,
        }.ToFrozenDictionary(StringComparer.Ordinal);

    public static readonly IReadOnlySet<string> DistrictIds = Districts.Select(item => item.Id).ToFrozenSet(StringComparer.Ordinal);

    public static readonly IReadOnlySet<string> VehicleIdSet = VehicleIds.ToFrozenSet(StringComparer.Ordinal);

    private static ZoneDefinition Zone(
        string id,
        string label,
        string[] aliases,
        int color,
        double happiness,
        double landValue,
        string kind = "DEVELOPMENT",
        string releaseScope = "MVP") =>
        new(id, label, Array.AsReadOnly(aliases), color, happiness, landValue, kind, releaseScope);
}
